Numpy support w/ python3crystax

I'm trying to get my app to compile with Python 3 using the Crystax NDK.  Everything is set up, and if I try to push my app to my device without specifying that numpy is needed, everything works fine.  However, numpy raises a few issues:
1. (easily fixed) The recipe needs to be modified such that python3crystax is supported.  Currently, it only claims python2 support, but that's overridden easily enough (`depends = ['python2']` -> `depends = [('python3crystax', 'python2)]` in the `__init__.py`).
2. The compiler claims it can't find -lcrystax:

``` sh
...
don't know how to compile Fortran code on platform 'posix'
C compiler: /usr/bin/ccache arm-linux-androideabi-gcc -DANDROID -mandroid -fomit-frame-pointer --sysroot /home/wwoods/walt/dev/kivy/crystax-ndk-10.3.2/platforms/android-19/arch-arm -DNDEBUG -g -fwrapv -O3 -Wall -Wstrict-prototypes -DANDROID -mandroid -fomit-frame-pointer --sysroot /home/wwoods/walt/dev/kivy/crystax-ndk-10.3.2/platforms/android-19/arch-arm -march=armv7-a -mfloat-abi=softfp -mfpu=vfp -mthumb -fPIC

compile options: '-Inumpy/core/src/private -Inumpy/core/src -Inumpy/core -Inumpy/core/src/npymath -Inumpy/core/src/multiarray -Inumpy/core/src/umath -Inumpy/core/src/npysort -Inumpy/core/include -I/home/wwoods/miniconda3/envs/kivy/include/python3.5m -c'
ccache: _configtest.c
/usr/bin/ccache arm-linux-androideabi-gcc -DANDROID -mandroid -fomit-frame-pointer --sysroot /home/wwoods/walt/dev/kivy/crystax-ndk-10.3.2/platforms/android-19/arch-arm _configtest.o -o _configtest
/home/wwoods/walt/dev/kivy/crystax-ndk-10.3.2/toolchains/arm-linux-androideabi-5/prebuilt/linux-x86_64/bin/../lib/gcc/arm-linux-androideabi/5.3/../../../../arm-linux-androideabi/bin/ld: error: cannot find -lcrystax
collect2: error: ld returned 1 exit status
/home/wwoods/walt/dev/kivy/crystax-ndk-10.3.2/toolchains/arm-linux-androideabi-5/prebuilt/linux-x86_64/bin/../lib/gcc/arm-linux-androideabi/5.3/../../../../arm-linux-androideabi/bin/ld: error: cannot find -lcrystax
collect2: error: ld returned 1 exit status
failure.
removing: _configtest.c _configtest.o
```

Everything except numpy works fine, so I think that it's just Numpy's recipe that needs a minor alteration.  Oddly though, specifying a LIBRARY_PATH or LDFLAGS does not have any effect.  In fact, if I add a `get_recipe_env` to the `NumpyRecipe` that prints out what `super().get_recipe_env(arch)` contains, it has LDFLAGS with the appropriate path on there.  Running the linker by hand with the given flags find the library fine (`ld $(LDFLAGS) -lcrystax --verbose`).  In the context of python-for-android though, the linker always fails to find libcrystax.a.

Any help appreciated, thanks.

Also been trying to get numpy working with crystax. It seems numpy's modified distutils doesn't use the env `LD`, `LDFLAGS`, or `CFLAGS`. You can get around the `-lcrystax` error above by adding the `-Lpath/to/crystax/sources/libs/<arch>`  to the `env['CC']`. 
After doing that it is able to pass the toolchain check but fails  when "Python.h" is included.
which it then complains as it uses the system python headers

```
compilation terminated.
failure.
removing: _configtest.c _configtest.o
Traceback (most recent call last):
  File "setup.py", line 392, in <module>
    setup_package()
  File "setup.py", line 384, in setup_package
    setup(**metadata)
  File "/home/user/.local/share/python-for-android/build/other_builds/numpy-python2crystax/x86/numpy/numpy/distutils/core.py", line 169, in setup
    return old_setup(**new_attr)
  File "/usr/lib/python2.7/distutils/core.py", line 151, in setup
    dist.run_commands()
  File "/usr/lib/python2.7/distutils/dist.py", line 953, in run_commands
    self.run_command(cmd)
  File "/usr/lib/python2.7/distutils/dist.py", line 972, in run_command
    cmd_obj.run()
  File "/home/user/.local/share/python-for-android/build/other_builds/numpy-python2crystax/x86/numpy/numpy/distutils/command/build_ext.py", line 82, in run
    self.run_command('build_src')
  File "/usr/lib/python2.7/distutils/cmd.py", line 326, in run_command
    self.distribution.run_command(command)
  File "/usr/lib/python2.7/distutils/dist.py", line 972, in run_command
    cmd_obj.run()
  File "/home/user/.local/share/python-for-android/build/other_builds/numpy-python2crystax/x86/numpy/numpy/distutils/command/build_src.py", line 148, in run
    self.build_sources()
  File "/home/user/.local/share/python-for-android/build/other_builds/numpy-python2crystax/x86/numpy/numpy/distutils/command/build_src.py", line 165, in build_sources
    self.build_extension_sources(ext)
  File "/home/user/.local/share/python-for-android/build/other_builds/numpy-python2crystax/x86/numpy/numpy/distutils/command/build_src.py", line 324, in build_extension_sources
    sources = self.generate_sources(sources, ext)
  File "/home/user/.local/share/python-for-android/build/other_builds/numpy-python2crystax/x86/numpy/numpy/distutils/command/build_src.py", line 377, in generate_sources
    source = func(extension, build_dir)
  File "numpy/core/setup.py", line 421, in generate_config_h
    moredefs, ignored = cocache.check_types(config_cmd, ext, build_dir)
  File "numpy/core/setup.py", line 47, in check_types
    out = check_types(*a, **kw)
  File "numpy/core/setup.py", line 280, in check_types
    "install {0}-dev|{0}-devel.".format(python))
SystemError: Cannot compile 'Python.h'. Perhaps you need to install python-dev|python-devel.

```
And you can fix that by adding an `-Ipath/to/crystax/python/includes/python` the same way.

However it then complains about missing symbols (as my current understanding of this is that numpy is omehow trying to use the debug version).

```
numpy/core/src/dummymodule.c:46: error: undefined reference to 'Py_InitModule4'
numpy/core/src/dummymodule.c:46: error: undefined reference to 'Py_InitModule4'

```
Maybe that helps... numpy with cryxtax would be nice.

Other important point:
I followed instructions to ignore the blas lapack atlas inclusion, which is possible but we loose a lot of speed with algebra computation (see https://docs.scipy.org/doc/numpy-1.9.2/user/install.html and https://stackoverflow.com/questions/32228967/compile-numpy-without-intel-mkl-blas-atlas-lapack). It should work but I still get the same last error and I think it's this one that halts everything:
```
C compiler: /usr/bin/ccache arm-linux-androideabi-gcc -DANDROID -mandroid -fomit-frame-pointer -D__ANDROID_API__=19 -I/media/axel/reste/.buildozer/crystax-ndk-10.3.2/platforms/android-19/arch-arm -isysroot /media/axel/reste/.buildozer/crystax-ndk-10.3.2/platforms/android-19/arch-arm -I/media/axel/reste/buildozer/.buildozer/android/platform/build/build/python-installs/medmanip/include/python3.5 -L/media/axel/reste/.buildozer/crystax-ndk-10.3.2/platforms/android-19/arch-arm/usr/lib/ --sysroot=/media/axel/reste/.buildozer/crystax-ndk-10.3.2/platforms/android-19/arch-arm -DNDEBUG -g -fwrapv -O2 -Wall -Wstrict-prototypes -DANDROID -mandroid -fomit-frame-pointer -D__ANDROID_API__=19 -I/media/axel/reste/.buildozer/crystax-ndk-10.3.2/platforms/android-19/arch-arm -isysroot /media/axel/reste/.buildozer/crystax-ndk-10.3.2/platforms/android-19/arch-arm -I/media/axel/reste/buildozer/.buildozer/android/platform/build/build/python-installs/medmanip/include/python3.5 -march=armv7-a -mfloat-abi=softfp -mfpu=vfp -mthumb -Wdate-time -D_FORTIFY_SOURCE=2 -fPIC

compile options: '-Inumpy/core/src/private -Inumpy/core/src -Inumpy/core -Inumpy/core/src/npymath -Inumpy/core/src/multiarray -Inumpy/core/src/umath -Inumpy/core/src/npysort -Inumpy/core/include -I/usr/include/python3.5m -c'
ccache: _configtest.c
/usr/bin/ccache arm-linux-androideabi-gcc -DANDROID -mandroid -fomit-frame-pointer -D__ANDROID_API__=19 -I/media/axel/reste/.buildozer/crystax-ndk-10.3.2/platforms/android-19/arch-arm -isysroot /media/axel/reste/.buildozer/crystax-ndk-10.3.2/platforms/android-19/arch-arm -I/media/axel/reste/buildozer/.buildozer/android/platform/build/build/python-installs/medmanip/include/python3.5 -L/media/axel/reste/.buildozer/crystax-ndk-10.3.2/platforms/android-19/arch-arm/usr/lib/ --sysroot=/media/axel/reste/.buildozer/crystax-ndk-10.3.2/platforms/android-19/arch-arm _configtest.o -o _configtest
/media/axel/reste/.buildozer/crystax-ndk-10.3.2/toolchains/arm-linux-androideabi-5/prebuilt/linux-x86_64/bin/../lib/gcc/arm-linux-androideabi/5.3/../../../../arm-linux-androideabi/bin/ld: error: cannot find -lcrystax
collect2: error: ld returned 1 exit status
/media/axel/reste/.buildozer/crystax-ndk-10.3.2/toolchains/arm-linux-androideabi-5/prebuilt/linux-x86_64/bin/../lib/gcc/arm-linux-androideabi/5.3/../../../../arm-linux-androideabi/bin/ld: error: cannot find -lcrystax
collect2: error: ld returned 1 exit status
failure.
```
lcrystax is missing!
In #1074 it was resolved by adding the libraries (seems like it works with python2crystax) but with the exact same path now it gives the same error they had before. So I keep thinking there is a problem with python3crystax, I would prefer there isn't because I tried to pick the content or android-19/arch-arm in there latest version (b1052) and the .a libs are still empty...

N.B: The warnings saying lapack, blas and atlas are missing don't fail the compilation. The failure is due to crystax missing.


I resolved this error by editing the flags in numpy recipe adding `-L{ctx.ndk_dir}/sources/crystax/libs/{arch.arch}` the the compilation started to work and again failed with an error over python.h.
After searching for some time I found that it was trying to compile my python3.5 installed in my computer rather than crystax's one.
I don't know why it did that but I edited the flags once more and it resolved this problem `-I{ctx.ndk_dir}/sources/python/{py_ver}/include/python/`.
By the way, after looking at the flags there https://github.com/frmdstryr/p4a-numpy/blob/master/p4a_numpy/__init__.py (it is what was used to edit the broken recipe of numpy for the py2.7 version last January or December), I saw that it had the correct flags to avoid this so I copy pasted all their flags.

Now the new problem is when building the extensions:
```
ccache: numpy/core/src/dummymodule.c
[.........]
creating build/lib.linux-x86_64-3.5
creating build/lib.linux-x86_64-3.5/numpy
creating build/lib.linux-x86_64-3.5/numpy/core
```
It's trying to build a library but I get this error:
`........ crystax-ndk-10.3.2/platforms/android-19/arch-arm/usr/lib/../lib/crtbegin_so.o: Relocations in generic ELF (EM: 40)`
After searching on internet I found that it's either due to using the wrong crtbegin, by wrong I mean one that has not been compiled with the same compilator, this one is in arch-arm so I hope it's good, or it could be because the compilator doesn't know how to use it and it's resolved with flags (good, until know everything has been resolved with flags).

Now I'm stuck, I don't know what to do because CC and CXX already have what I intended to put : `/usr/bin/ccache arm-linux-androideabi-gcc` (and g++ respectively) 
Unfortunately I don't have anymore time to work on this but I hope my explanations will save a lot of time to the next person trying to resolve this problem.

Here the last error giving me a headache, I'm not used to compilation problems but isn't it strange that it uses linux x86_64 to compile for arm? Until now the commands visible in the console started with `C compiler: /usr/bin/ccache arm-linux-androideabi-gcc -DANDROID -mandroid -fomit-frame-poi`

`error: Command "x86_64-linux-gnu-gcc -pthread -shared -Wl,-O1 -Wl,-Bsymbolic-functions -Wl,-Bsymbolic-functions -Wl,-z,relro --sysroot /media/axel/reste/.buildozer/crystax-ndk-10.3.2/platforms/android-19/arch-arm -lm -L/media/axel/reste/buildozer/.buildozer/android/platform/build/build/libs_collections/medmanip/armeabi-v7a -L/media/axel/reste/.buildozer/crystax-ndk-10.3.2/sources/crystax/libs/armeabi-v7a -lcrystax -DANDROID -mandroid -fomit-frame-pointer -D__ANDROID_API__=19 -I/media/axel/reste/.buildozer/crystax-ndk-10.3.2/platforms/android-19/arch-arm -isysroot /media/axel/reste/.buildozer/crystax-ndk-10.3.2/platforms/android-19/arch-arm -I/media/axel/reste/buildozer/.buildozer/android/platform/build/build/python-installs/medmanip/include/python3.5 -march=armv7-a -Wdate-time -D_FORTIFY_SOURCE=2 build/temp.linux-x86_64-3.5/numpy/core/src/dummymodule.o -Lbuild/temp.linux-x86_64-3.5 -o build/lib.linux-x86_64-3.5/numpy/core/_dummy.cpython-35m-x86_64-linux-gnu.so" failed with exit status 1`

--- a/pythonforandroid/recipes/numpy/__init__.py
+++ b/pythonforandroid/recipes/numpy/__init__.py
@@ -1,48 +1,48 @@
 from pythonforandroid.recipe import CompiledComponentsPythonRecipe
 from pythonforandroid.toolchain import warning
+from os.path import join
 
 
 class NumpyRecipe(CompiledComponentsPythonRecipe):
 
-    version = '1.9.2'
-    url = 'https://pypi.python.org/packages/source/n/numpy/numpy-{version}.tar.gz'
-    site_packages_name= 'numpy'
+    version = '1.15.1'
+    url = 'https://pypi.python.org/packages/source/n/numpy/numpy-{version}.zip'
+    site_packages_name = 'numpy'
 
-    depends = ['python2']
+    depends = [('python2', 'python3crystax')]
 
-    patches = ['patches/fix-numpy.patch',
-               'patches/prevent_libs_check.patch',
-               'patches/ar.patch',
-               'patches/lib.patch']
+    patches = [
+        join('patches', 'fix-numpy.patch'),
+        join('patches', 'prevent_libs_check.patch'),
+        join('patches', 'ar.patch'),
+        join('patches', 'lib.patch'),
+        join('patches', 'python2-fixes.patch')
+    ]
 
     def get_recipe_env(self, arch):
-        """ looks like numpy has no proper -L flags. Code copied and adapted from
-            https://github.com/frmdstryr/p4a-numpy/
-        """
-
         env = super(NumpyRecipe, self).get_recipe_env(arch)
-        #: Hack add path L to crystax as a CFLAG
 
-        py_ver = '3.5'
-        if {'python2crystax', 'python2'} & set(self.ctx.recipe_build_order):
-            py_ver = '2.7'
+        flags = " -L{} --sysroot={}".format(
+            join(self.ctx.ndk_platform, 'usr', 'lib'),
+            self.ctx.ndk_platform
+        )
 
-        py_so = '2.7' if py_ver == '2.7' else '3.5m'
+        if self.ctx.ndk == 'crystax':
+            py_ver = self.ctx.python_recipe.version[0:3]
+            src_dir = join(self.ctx.ndk_dir, 'sources')
+            py_inc_dir = join(src_dir, 'python', py_ver, 'include', 'python')
+            py_lib_dir = join(src_dir, 'python', py_ver, 'libs', arch.arch)
+            cry_inc_dir = join(src_dir, 'crystax', 'include')
+            cry_lib_dir = join(src_dir, 'crystax', 'libs', arch.arch)
+            flags += ' -I{}'.format(py_inc_dir)
+            flags += ' -L{} -lpython{}m'.format(py_lib_dir, py_ver)
+            flags += " -I{}".format(cry_inc_dir)
+            flags += " -L{}".format(cry_lib_dir)
 
-        api_ver = self.ctx.android_api
-
-        platform = 'arm' if 'arm' in arch.arch else arch.arch
-        #: Not sure why but we have to inject these into the CC and LD env's for it to
-        #: use the correct arguments.
-        flags = " -L{ctx.ndk_dir}/platforms/android-{api_ver}/arch-{platform}/usr/lib/" \
-                " --sysroot={ctx.ndk_dir}/platforms/android-{api_ver}/arch-{platform}" \
-            .format(ctx=self.ctx, arch=arch, platform=platform, api_ver=api_ver,
-                    py_so=py_so, py_ver=py_ver)
         if flags not in env['CC']:
             env['CC'] += flags
         if flags not in env['LD']:
             env['LD'] += flags + ' -shared'
-
         return env
 
     def prebuild_arch(self, arch):

--- a/pythonforandroid/recipes/numpy/patches/ar.patch
+++ b/pythonforandroid/recipes/numpy/patches/ar.patch
@@ -1,11 +1,44 @@
---- a/numpy/distutils/unixccompiler.py	2015-02-01 17:38:21.000000000 +0100
-+++ b/numpy/distutils/unixccompiler.py	2015-07-08 17:21:05.742468485 +0200
-@@ -82,6 +82,8 @@
-             pass
-         self.mkpath(os.path.dirname(output_filename))
-         tmp_objects = objects + self.objects
-+        from os import environ
-+        self.archiver[0] = 'arm-linux-androideabi-ar'
+diff --git a/numpy/core/code_generators/generate_umath.py b/numpy/core/code_generators/generate_umath.py
+index 632bcb4..c1e0dd5 100644
+--- a/numpy/core/code_generators/generate_umath.py
++++ b/numpy/core/code_generators/generate_umath.py
+@@ -970,6 +970,7 @@ def make_arrays(funcdict):
+                 funclist.append('%s_%s' % (tname, name))
+                 if t.simd is not None:
+                     for vt in t.simd:
++                        continue
+                         code2list.append(textwrap.dedent("""\
+                         #ifdef HAVE_ATTRIBUTE_TARGET_{ISA}
+                         if (npy_cpu_supports("{isa}")) {{
+diff --git a/numpy/distutils/ccompiler.py b/numpy/distutils/ccompiler.py
+index b03fb96..f9e6cd0 100644
+--- a/numpy/distutils/ccompiler.py
++++ b/numpy/distutils/ccompiler.py
+@@ -275,6 +275,7 @@ def CCompiler_compile(self, sources, output_dir=None, macros=None,
+             self._setup_compile(output_dir, macros, include_dirs, sources,
+                                 depends, extra_postargs)
+     cc_args = self._get_cc_args(pp_opts, debug, extra_preargs)
++    cc_args += os.environ['CFLAGS'].split()
+     display = "compile options: '%s'" % (' '.join(cc_args))
+     if extra_postargs:
+         display += "\nextra options: '%s'" % (' '.join(extra_postargs))
+diff --git a/numpy/distutils/unixccompiler.py b/numpy/distutils/unixccompiler.py
+index 11b2cce..f6dde79 100644
+--- a/numpy/distutils/unixccompiler.py
++++ b/numpy/distutils/unixccompiler.py
+@@ -54,6 +54,7 @@ def UnixCCompiler__compile(self, obj, src, ext, cc_args, extra_postargs, pp_opts
+         deps = []
+ 
+     try:
++        self.linker_so = [os.environ['LD']+" "+os.environ['LDFLAGS']]
+         self.spawn(self.compiler_so + cc_args + [src, '-o', obj] + deps +
+                    extra_postargs, display = display)
+     except DistutilsExecError:
+@@ -111,6 +112,7 @@ def UnixCCompiler_create_static_lib(self, objects, output_libname,
          while tmp_objects:
              objects = tmp_objects[:50]
              tmp_objects = tmp_objects[50:]
++            self.archiver[0] = 'arm-linux-androideabi-ar'
+             display = '%s: adding %d object files to %s' % (
+                            os.path.basename(self.archiver[0]),
+                            len(objects), output_filename)

--- a/pythonforandroid/recipes/numpy/patches/fix-numpy.patch
+++ b/pythonforandroid/recipes/numpy/patches/fix-numpy.patch
@@ -1,49 +1,17 @@
-diff --git a/numpy/core/src/multiarray/numpyos.c b/numpy/core/src/multiarray/numpyos.c
-index 44b32f4..378e199 100644
---- a/numpy/core/src/multiarray/numpyos.c
-+++ b/numpy/core/src/multiarray/numpyos.c
-@@ -165,8 +165,7 @@ ensure_decimal_point(char* buffer, size_t buf_size)
- static void
- change_decimal_from_locale_to_dot(char* buffer)
- {
--    struct lconv *locale_data = localeconv();
--    const char *decimal_point = locale_data->decimal_point;
-+    const char *decimal_point = ".";
- 
-     if (decimal_point[0] != '.' || decimal_point[1] != 0) {
-         size_t decimal_point_len = strlen(decimal_point);
-@@ -448,8 +447,7 @@ NumPyOS_ascii_strtod_plain(const char *s, char** endptr)
- NPY_NO_EXPORT double
- NumPyOS_ascii_strtod(const char *s, char** endptr)
- {
--    struct lconv *locale_data = localeconv();
--    const char *decimal_point = locale_data->decimal_point;
-+    const char *decimal_point = ".";
-     size_t decimal_point_len = strlen(decimal_point);
- 
-     char buffer[FLOAT_FORMATBUFLEN+1];
-diff --git a/numpy/core/src/private/npy_config.h b/numpy/core/src/private/npy_config.h
-index f768c90..4e5d168 100644
---- a/numpy/core/src/private/npy_config.h
-+++ b/numpy/core/src/private/npy_config.h
-@@ -41,4 +41,10 @@
- #undef HAVE_ATAN2
- #endif
- 
-+/* Android only */
-+#ifdef ANDROID
-+#undef HAVE_LDEXPL
-+#undef HAVE_FREXPL
-+#endif
-+
- #endif
 diff --git a/numpy/testing/__init__.py b/numpy/testing/__init__.py
-index 258cbe9..ce4e0eb 100644
+index a7c8593..007ce26 100644
 --- a/numpy/testing/__init__.py
 +++ b/numpy/testing/__init__.py
-@@ -1,16 +1,7 @@
+@@ -1,22 +1,8 @@
 -"""Common test support for all numpy test scripts.
--
++# fake tester, android don't have unittest
++class Tester(object):
++    def test(self, *args, **kwargs):
++        pass
++    def bench(self, *args, **kwargs):
++        pass
++test = Tester().test
+ 
 -This single module should provide all the common functionality for numpy tests
 -in a single location, so that test scripts can just import it and work right
 -away.
@@ -53,14 +21,14 @@ index 258cbe9..ce4e0eb 100644
 -
 -from unittest import TestCase
 -
--from . import decorators as dec
--from .utils import *
--from .nosetester import NoseTester as Tester
--from .nosetester import run_module_suite
-+# fake tester, android don't have unittest
-+class Tester(object):
-+    def test(self, *args, **kwargs):
-+        pass
-+    def bench(self, *args, **kwargs):
-+        pass
- test = Tester().test
+-from ._private.utils import *
+-from ._private import decorators as dec
+-from ._private.nosetester import (
+-    run_module_suite, NoseTester as Tester
+-    )
+-
+-__all__ = _private.utils.__all__ + ['TestCase', 'run_module_suite']
+-
+-from ._private.pytesttester import PytestTester
+-test = PytestTester(__name__)
+-del PytestTester

--- a/pythonforandroid/recipes/numpy/patches/lib.patch
+++ b/pythonforandroid/recipes/numpy/patches/lib.patch
@@ -1,39 +1,43 @@
---- a/numpy/linalg/setup.py	2015-07-09 14:15:59.850853336 +0200
-+++ b/numpy/linalg/setup.py	2015-07-09 14:21:59.403889000 +0200
-@@ -37,7 +37,8 @@
-     config.add_extension('lapack_lite',
-                          sources = [get_lapack_lite_sources],
-                          depends = ['lapack_litemodule.c'] + lapack_lite_src,
--                         extra_info = lapack_info
-+                         extra_info = lapack_info,
-+                         libraries = ['m'],
-                          )
- 
-     # umath_linalg module
-@@ -46,7 +47,7 @@
-                          sources = [get_lapack_lite_sources],
-                          depends =  ['umath_linalg.c.src'] + lapack_lite_src,
-                          extra_info = lapack_info,
--                         libraries = ['npymath'],
-+                         libraries = ['npymath','m'],
-                          )
- 
-     return config
---- a/numpy/fft/setup.py	2015-07-09 14:35:22.299888028 +0200
-+++ b/numpy/fft/setup.py	2015-07-09 14:33:54.858392578 +0200
-@@ -9,7 +9,8 @@
+diff --git a/numpy/fft/setup.py b/numpy/fft/setup.py
+index cd99a82d7..e614ecd07 100644
+--- a/numpy/fft/setup.py
++++ b/numpy/fft/setup.py
+@@ -9,7 +9,8 @@ def configuration(parent_package='',top_path=None):
  
      # Configure fftpack_lite
      config.add_extension('fftpack_lite',
 -                         sources=['fftpack_litemodule.c', 'fftpack.c']
 +                         sources=['fftpack_litemodule.c', 'fftpack.c'],
-+                         libraries = ['m']
++                         libraries=['m']
                           )
  
+     return config
+diff --git a/numpy/linalg/setup.py b/numpy/linalg/setup.py
+index 66c07c9e1..d34bd930a 100644
+--- a/numpy/linalg/setup.py
++++ b/numpy/linalg/setup.py
+@@ -43,6 +43,7 @@ def configuration(parent_package='', top_path=None):
+         sources=['lapack_litemodule.c', get_lapack_lite_sources],
+         depends=['lapack_lite/f2c.h'],
+         extra_info=lapack_info,
++        libraries=['m'],
+     )
  
---- a/numpy/random/setup.orig.py	2015-07-09 14:44:41.105174826 +0200
-+++ b/numpy/random/setup.py	2015-07-09 14:46:08.592679877 +0200
-@@ -38,7 +38,7 @@
+     # umath_linalg module
+@@ -51,7 +52,7 @@ def configuration(parent_package='', top_path=None):
+         sources=['umath_linalg.c.src', get_lapack_lite_sources],
+         depends=['lapack_lite/f2c.h'],
+         extra_info=lapack_info,
+-        libraries=['npymath'],
++        libraries=['npymath', 'm'],
+     )
+     return config
+ 
+diff --git a/numpy/random/setup.py b/numpy/random/setup.py
+index 3f3b773a4..c1db9f783 100644
+--- a/numpy/random/setup.py
++++ b/numpy/random/setup.py
+@@ -40,7 +40,7 @@ def configuration(parent_package='',top_path=None):
      if needs_mingw_ftime_workaround():
          defs.append(("NPY_NEEDS_MINGW_TIME_WORKAROUND", None))
  

--- a/pythonforandroid/recipes/numpy/patches/prevent_libs_check.patch
+++ b/pythonforandroid/recipes/numpy/patches/prevent_libs_check.patch
@@ -1,8 +1,8 @@
 diff --git a/numpy/distutils/system_info.py b/numpy/distutils/system_info.py
-index a050430..471e958 100644
+index bea120cf9..a448a83fc 100644
 --- a/numpy/distutils/system_info.py
 +++ b/numpy/distutils/system_info.py
-@@ -610,6 +610,7 @@ class system_info:
+@@ -719,6 +719,7 @@ class system_info(object):
          return self.get_paths(self.section, key)
  
      def get_libs(self, key, default):

--- a/pythonforandroid/recipes/numpy/patches/python2-fixes.patch
+++ b/pythonforandroid/recipes/numpy/patches/python2-fixes.patch
@@ -1,0 +1,69 @@
+diff --git a/numpy/core/src/multiarray/common.c b/numpy/core/src/multiarray/common.c
+index c70f852..695efd5 100644
+--- a/numpy/core/src/multiarray/common.c
++++ b/numpy/core/src/multiarray/common.c
+@@ -852,3 +852,12 @@ _may_have_objects(PyArray_Descr *dtype)
+     return (PyDataType_HASFIELDS(base) ||
+             PyDataType_FLAGCHK(base, NPY_ITEM_HASOBJECT) );
+ }
++
++/*
++ * Dummy to fix android NDK problem with missing reference.
++ */
++void *
++__emutls_get_address(struct __emutls_object *obj)
++{
++    return NULL;
++}
+diff --git a/numpy/distutils/exec_command.py b/numpy/distutils/exec_command.py
+index 8118e2f..b586442 100644
+--- a/numpy/distutils/exec_command.py
++++ b/numpy/distutils/exec_command.py
+@@ -260,7 +260,7 @@ def _exec_command(command, use_shell=None, use_tee = None, **env):
+         return 127, ''
+ 
+     text, err = proc.communicate()
+-    text = text.decode(locale.getpreferredencoding(False),
++    text = text.decode('UTF-8',
+                        errors='replace')
+ 
+     text = text.replace('\r\n', '\n')
+diff --git a/numpy/distutils/misc_util.py b/numpy/distutils/misc_util.py
+index f2d677a..758b1ed 100644
+--- a/numpy/distutils/misc_util.py
++++ b/numpy/distutils/misc_util.py
+@@ -9,7 +9,6 @@ import atexit
+ import tempfile
+ import subprocess
+ import shutil
+-import multiprocessing
+ 
+ import distutils
+ from distutils.errors import DistutilsError
+@@ -93,10 +92,7 @@ def get_num_build_jobs():
+ 
+     """
+     from numpy.distutils.core import get_distribution
+-    try:
+-        cpu_count = len(os.sched_getaffinity(0))
+-    except AttributeError:
+-        cpu_count = multiprocessing.cpu_count()
++    cpu_count = 1
+     envjobs = int(os.environ.get("NPY_NUM_BUILD_JOBS", cpu_count))
+     dist = get_distribution()
+     # may be None during configuration
+diff --git a/setup.py b/setup.py
+index fed178e..b0266eb 100755
+--- a/setup.py
++++ b/setup.py
+@@ -377,9 +377,8 @@ def setup_package():
+         # Raise errors for unsupported commands, improve help output, etc.
+         run_build = parse_setuppy_commands()
+ 
+-    from setuptools import setup
++    from numpy.distutils.core import setup
+     if run_build:
+-        from numpy.distutils.core import setup
+         cwd = os.path.abspath(os.path.dirname(__file__))
+         if not os.path.exists(os.path.join(cwd, 'PKG-INFO')):
+             # Generate Cython sources, unless building from source release

